http-clients/src/main/java/org/triplea/http/client/GenericWebSocketClient.java

```diff
@@ -0,0 +1,147 @@
+          public void onOpen(final ServerHandshake serverHandshake) {}
+
+          @Override
+          public void onMessage(final String message) {
+            notifyListeners(message, incomingMessageType);
+          }
+
+          @Override
+          public void onClose(final int code, final String reason, final boolean remote) {
+            connectionLostListener.accept(reason);
+          }
+
+          @Override
+          public void onError(final Exception ex) {
+            connectionLostListener.accept(ex.getMessage());
+          }
+        };
+    client.connect();
+  }
+
+  private void notifyListeners(final String message, final Class<IncomingT> messageType) {
+    final IncomingT converted = gson.fromJson(message, messageType);
+    messageListeners.forEach(listener -> listener.accept(converted));
+  }
+
+  /**
+   * Non-blocking send of a message to the server. Implementation note: data is sent as a JSON
+   * string, this method handles conversion of the parameter object to JSON.
+   *
+   * @param message The data object to send to the server.
+   */
+  public void send(final OutgoingT message) {
+    final String json = gson.toJson(message);
+    final Future<?> result = threadPool.submit(() -> sendMessage(json));
+
+    // wait for result on a new thread so we do not block the current one
+    // this is only done so we can report any errors that have occurred.
+    new Thread(
+            () -> {
+              try {
+                result.get();
+              } catch (final InterruptedException e) {
+                Thread.currentThread().interrupt();
+              } catch (final ExecutionException e) {
+                log.log(Level.SEVERE, "Failed sending message: " + e.getMessage(), e);
+              }
+            })
+        .start();
+    ;
+  }
+
+  private void sendMessage(final String message) {
+    Preconditions.checkState(!closed);
+    if (!connectionIsOpen) {
+      waitForConnection();
+    }
+
+    if (connectionIsOpen) {
+      // TODO: make sure client connection is open 'client.isOpen()' or throw
+      client.send(message);
+    }
+  }
+
+  private void waitForConnection() {
+    final long startWait = System.currentTimeMillis();
+    while (!client.isOpen()) {
+      try {
+        Thread.sleep(10L);
+        final long elapsed = System.currentTimeMillis() - startWait;
+        if (elapsed > CONNECT_TIMEOUT_MILLIS) {
+          return;
+        }
+      } catch (final InterruptedException e) {
+        Thread.currentThread().interrupt();
+        return;
+      }
+    }
+    connectionIsOpen = true;
+  }
+
+  /**
+   * Registers a callback that is invoked when messages are received. Multiple message listeners may
+   * be added.
+   */
+  public void addMessageListener(final Consumer<IncomingT> listener) {
+    messageListeners.add(listener);
+  }
+
+  /** Non-blocking close of the websocket connection. */
+  public void close() {
+    closed = true;
+    connectionIsOpen = false;
```

Comment:
This level of detail is a bit rough, so grain of salt there.

The two are slightly different:
- closed == we have closed the connection, sending messages is illegal operation
- connectionIsOpen == we have requested for the connection to become open, it takes about 200ms for that operation to complete.  The flag is used to avoid future wait loops to check if the connection is open. I suspect it will be just as well to do the wait-loop each time given it'll return right away if the connection is open.

Indeed, grain of salt with this change here, any class that is not yet unit tested is prototype quality. Overall this branch is a very advanced prototype.